### DISCUSSION TITLE
Vergleich Berechnungsskripte

### DISCUSSION THREAD

**Discussion by pheagan** [atticus-sullivan/sim-ayto#3] [General]
Hey, hier der Erdbeerkäse Alex,
hab mein Skript mal etwas aufgeräumt und auch auf Github hochgeladen:
https://github.com/pheagan/AYTO-Berechnung
Hat bisher nur den beschriebenen Funktionsumfang und den etwas aufwändigeren Input, weil ich da bisher nicht die Notwendigkeit gesehen habe (und Motivation habe) das umzubauen.
Bin bisher nicht dazu gekommen mir anzuschauen, wie du das bei dir laufen lässt, hab wie gesagt eigentlich nichts mit Programmieren zu tun^^ Aber kannst auch gern mal in mein Skript reinschauen.
LG

**Comment by atticus-sullivan** [comment]
Oh hi,

> den etwas aufwändigeren Input, weil ich da bisher nicht die Notwendigkeit gesehen habe (und Motivation habe) das umzubauen.

Muss ja nich, is ja nur n kleines Hobby Projekt das erstmal nich unbedingt für andere gedacht war.

> Aber kannst auch gern mal in mein Skript reinschauen.

Hab mal reingeschaut, über ein/zwei Sachen muss ich nochmal nachdenken aber sonst siehts schonmal interesant aus. Soll ich dir n kleines Code-Review machen oder hast du da kein Interesse? ^^

> wie du das bei dir laufen lässt,

Ja is halt Lua (ne Art do it yourself Sprache wo man das meiste selbst machen muss (zB permutationen), aber im Idealfall is da das meiste hinter Funktionen versteckt und sollte lesbar sein) kennst wahrscheinlich nich, aber eigentlich sollte man es halbwegs lesen können (zumindest so die Grundgedanken). Aber klar, mach wie du Lust und Zeit hast, is ja alles Freizeit!

**Reply by pheagan** [reply]
Du darfst gern ein Review machen, aber wär jetzt auch nicht nötig. Das einzige was ich da von Zeit zu Zeit optimiert hatte war hinsichtlich Zeit, wenn der Rest nicht optimal sauber geschrieben ist, ist mir das egal :D
Ich schau aktuell noch ob ich meinen Plan mit der interaktiven Tabelle umsetzen kann, danach geh ich vielleicht mal noch den Input an

**Reply by atticus-sullivan** [reply]
Ja mal schaun wenn ich Zeit hab, mach ich das mal (vielleicht kann ich ja n bissl cooles Python-Stuff unterbringen ^^, aber du sollst den Code natürlich schon noch verstehen. Aber python is ja idR sehr gut verständlich). 

Bzgl der Zeit hab ich gesehn, das is vA der Part, dass du die MatchBoxes auf die Möglichkeiten ohne Person 11 anwendest, oder?
Muss auch ehrlich sagen, ja meins läuft jetz ~9-10 Min (bei wieviel bist du eigentlich so ungefähr? (auch wenns nich vergleichbar is wegen Hardware etc)) aber auf so schnell, dass es "praktisch" ist bekommt mans bei der Menge glaub nich runter, und ab ner gewissen Zeit is es mir egal obs jetz (n bissl, zumindest glaub ich, dass es nur n bissl is) länger braucht oder nicht.
[Und ich glaub ich bleib da bei meinem unoptimierten, auch aus dem Grund, da ich auf die Art und Weise ne Art Chronologie hab (welcher Constraint hat wieviele Matchings rausgeworfen) -> die alten Tabellen (und evtl Bäume (wenn auch unsortiert von der Ebene her), aber da muss ich nochmal drüber nachdenken) erstellen kann]
Der Punkt wo wir beide halt total viel Potential liegen lassen ist, dass man das eigentlich wunderbar parallelisieren kann (sollte auf maximal `1/<anzahl der Kerne>` (nur als untere Schranke, der Wert is natürlich nicht erreichbar) sinken). In Lua is Parallelisierung n Albtraum von dem was ich bisher mal rumprobiert hab, aber in Python müsste das eigentlich gehn (`boolean_index.py` auf https://towardsdatascience.com/speeding-up-python-code-fast-filtering-and-slow-loops-8e11a09a9c2f sieht ganz spannend aus, braucht aber halt `numpy` als neue library. Nur falls du Bock hast die Zeit noch n bissl zu drücken)

Interaktive Tabelle is halt sone Sache, wenn man die Anzahl der Kombinationen constraint, könnte es sogar halbwegs gehn von der Zeit zum neu rechnen (die Möglichkeiten werden ja dann doch eigentlich recht schnell weniger). Sachen abwählen is halt aufwändig, weil die Matchings hat man ja eigentlich nich mehr die man jetzt hinzufügen müsste (meine Idee wär bisher (hab mir mal Gedanken gemacht, aber mir is da glaub die Hürde der GUI zu groß ^^) sich ne Art Snapshot speichern was da is ab dem Zeitpunkt wo der User anfängt, beim Abwählen dahin zurücksetzen und das noch angewählte dadrauf nochmal anwenden)

**Reply by pheagan** [reply]
Genau, erstmal alle Kombinationen rausschmeißen in der 10x10 Kombination, wobei hier alle Infos bezüglich der +11 Person ignoriert werden (daher auch keine MatchNights weil das da komplizierter wäre) und anschließend erst in 10x11 alle Informationen.
Bei mir hängt die Geschwindigkeit natürlich stark davon ab, wie viele Kombinationen ich im ersten Schritt rausschmeißen kann und wie viele Kombinationen sonst rausfliegen. Habs mal an nem leeren Input getestet (alle 36Mio Möglichkeiten): 
Erstellen der 10x10Matrix und Checken auf Matchboxen ~ 3s,
Erstellen 10x11 und Prüfen auf MatchNights und Matchboxen ~40s
Berechnen der relativen Tabelle aus allen gespeicherten Möglichkeiten ~1min (habs jetzt geändert dass ich Kombinationen als 1D-Liste speicher und nicht mehr als 2D-Matrix, dann kann ich in der zweiten Dimension einfach alle Kombinationen abspeichern)

Als Vergleich aktuelle Staffel nach Nacht 7:
Erstellen der 10x10Matrix und Checken auf Matchboxen ~ 2s,
Erstellen 10x11 und Prüfen auf MatchNights und Matchboxen ~1s
Berechnen der relativen Tabelle aus allen gespeicherten Möglichkeiten im ms-Bereich

Wichtig: Das sind die Zeiten, wenn ich das aus dem Explorer direkt laufen lass. Wenn ichs aus Visual Studio raus laufen lass dauert es viel länger (Minuten Bereich bis ~45s wenn nur wenige Kombinationen übrig sind).
Muss also gar nicht mehr so viel optimieren, aber ich war am Anfang, als ich auch die Permutationen noch über Schleifen erstellt hab wegen der 11x10 Asymetrie, bei 5-10min pro Durchlauf.

Interaktive Tabelle veröffentliche ich auch sobald ich soweit bin. Mein Plan: Ich speicher mir als Zwischenschritt alle noch möglichen Kombinationen in einer Liste ab und die Berechnung wenn ich zusätzlich Matches festlege basiert dann nur noch auf der Liste der restlichen Matches (max. ein paa Tausend ab Folge 3 oder so)

**Reply by atticus-sullivan** [reply]
Oh ok, dann hätte das auf den 10x10 Möglichkeiten arbeiten doch seinen Reiz. Weil genau die 5-10 min hab ich ja dann quasi auch derzeit.
Ja gut ka was Visual Studio da noch alles macht, aber wenns ausm Explorer raus so geht, sollte das ja ne echte Zeit sein.

> Interaktive Tabelle

Genau das wäre glaub auch mein Ansatz (und genau die Liste hast du ja glaub derzeit schon)

**Reply by atticus-sullivan** [reply]
>  dann hätte das auf den 10x10 Möglichkeiten arbeiten doch seinen Reiz. Weil genau die 5-10 min hab ich ja dann quasi auch derzeit.

Danke für den Hinweis, dass du ohne die Optimierung auch die 10 min hast und ohne deutlich weniger (weniger als ich multithreading erwarten würde). Hab wieder mal unterschätzt wie viel da dann doch noch dazu kommt. Jetz läuft meins auch in ~1 Minute.
[Ich merk mir einfach wann Person 11 dazu kommt (das hab ich ja quasi eh in meiner Config), wend bis dahin alles normal an (n bissl rumrechnen beim tracken wo wieviel rausgeflogen ist), füg dann die neuen Möglichkeiten hinzu und wend dann den Rest an (so kann ich die Chronologie behalten)].
Ohne den Hinweis hätt ich tatsächlich über den Punkt nimmer nachgedacht.

**Comment by atticus-sullivan** [comment]
Hab dir mal [hier](https://github.com/atticus-sullivan/tmp/pull/1#pullrequestreview-1120403832) n bissl was zu deinem Code geschrieben (is aber wirklich nur als kleine Anregung gedacht)

**Comment by atticus-sullivan** [comment]
Note am Rande, nach der MN (aufgelöst) in Episode 18 wissen sie das Ergebnis (theoretisch). Hast du das auch (immernoch dabei mein Script gegen zu checken xD)?

**Reply by pheagan** [reply]
Jap, da komm ich auch zu ;)
Poste das jetzt auch noch auf Patreon mit fettem Spoiler Hinweis^^

**Reply by atticus-sullivan** [reply]
Top:+1
Ja finds ganz lustig zu sehn wie sie rätseln für/bei der MB wenn mans schon weiß xD

**Reply by pheagan** [reply]
Und wieder zeigt sich: Die Mathematik gewinnt einfach gegen "das Herz" xD
Egal was Sophia Thomalla da sagt, immer mit Taktik rangehen!

**Reply by atticus-sullivan** [reply]
V.A. am Ende halt und vorher halt mal das Hirn einschalten und nich zweimal dasselbe probiern (aber dafür müssten ja alle an einem Strang ziehen... :see_no_evil:) xD

**Comment by pheagan** [comment]
Hey, ich glaub der Kommentar passt hier besser als unter dein Code Review:
Ich hatte endlich mal Zeit und konnte mich an die interaktive Tabelle setzen. Keine Aufforderung zum Review, der Fokus war wieder erstmal auf der Funktionalität, nicht auf optimalem/sauberem Coding. Lediglich der Hinweis, dass du da mal reinschauen und mit der neuen Tabelle rumspielen kannst ;)

**Reply by atticus-sullivan** [reply]
Oh cool, klingt interessant. Schau ich mir auf jeden Fall mal an. Das Interface sieht ja auch cool aus (bisher nur das Bild gesehn).
Code schau ich mir auch gern mal an, aber da brauch ich n bissl für (bis dahin is die Staffel wahrscheinlich schon durch ^^).

**Reply by pheagan** [reply]
Wie gesagt, Code Review muss nicht sein (außer das läuft bei dir jetzt aus irgend nem Grund nicht).
Das Ding läuft jetzt, hat den Umfang den ich will, die Staffel ist Berechnung eh schon durch, also wird das Skript erstmal wieder eingemottet bis Anfang nächsten Jahres zur neuen Staffel^^

**Reply by atticus-sullivan** [reply]
So es geht wieder los ^^. Hab mal n bissl was eingetragen für die neue Staffel:
[AYTO_Data.xlsx](https://github.com/atticus-sullivan/sim-ayto/files/12480929/AYTO_Data.xlsx) (evtl muss man da die Seiten aber auch nochmal switchen je nachdem ob und wenn ja wer noch nachkommt).

Aktuell läufts grad noch durch, aber bisher siehts glaub ganz gut aus.

EDIT: Hab anscheinend was beim zusätzliche Person ausbaun falsch gemacht:
```
Traceback (most recent call last):
  File "./AYTO_Berechnung.py", line 307, in <module>
    textBonusPerson = textBonusPerson + gruppeB[i] + ": " + str(percent) + "%, "
                      ~~~~~~~~~~~~~~~~^~~~~~~~~~~~
TypeError: can only concatenate str (not "float") to str
```

PS: Bin unter [s03r_tab.pdf](https://github.com/atticus-sullivan/sim-ayto/blob/build/s03r/s03r_tab.pdf) auch wieder dabei. (Und btw im Vergleich mit den anderen Staffeln läufts bisher in etwa durchschnittlich siehe [hier](https://github.com/atticus-sullivan/sim-ayto/blob/build/stats.pdf) (das Diagramm ganz rechts is bisher nich so aussagekräftig wegen der unterschiedlichen Personenanzahl))
